### PR TITLE
Add contract caching mechanism for acceptance tests

### LINKED ISSUE #6789: Implement contract caching mechanism
### Problem

Contracts are pretty expensive to create on main net and we are now creating the same contract(i.e. precompile contract) in multiple test suites(estimatePrecomple, precompile, call).

### Solution

Implement a caching mechanism that would get already created contract(singleton) similar to the `AccoutClient` and `TokenClient` and it would be used across all test suites.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by yiliev0**
**Description**: This PR introduces a caching mechanism to retrieve already created contracts, reducing the cost and redundancy of deploying the same contract in multiple test suites. 

**Related issue(s)**:

Fixes #6789

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java [hunk 1]
```diff
@@ -95,23 +104,30 @@ protected MirrorTransaction verifyMirrorTransactionsResponse(MirrorNodeClient mi
...
-                            .getSdkClient()
-                            .getAcceptanceTestProperties()
-                            .getFeatureProperties()
-                            .getMaxContractFunctionGas(),
-                    initialBalance == 0 ? null : Hbar.fromTinybars(initialBalance),
-                    null);
-            contractId = verifyCreateContractNetworkResponse();
-            return new DeployedContract(fileId, contractId, compiledSolidityArtifact);
-        }
+    protected synchronized DeployedContract getContract(ContractResource contractResource)
```
Comment (0xivanov) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java:107]: In Java, when you have a synchronized method in an abstract class and you create two different subclasses that inherit from this abstract class, the synchronization is based on the current object instance. Each subclass instance will have its own lock associated with it.
In this case we can modify the code like this, in order to synchronize on a shared object:
``` java
protected DeployedContract getContract(ContractResource contractResource) {
    synchronized (contractIdMap) {
        // the rest of the method
    }
}
```

Comment (0xivanov) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java:107]: As mentioned [here](https://stackoverflow.com/questions/14947723/is-concurrenthashmap-totally-safe), `ConcurrentHashMap` is not 100% thread safe when we do sequence of get and put operations. That is why we need to make the whole method atomic IMO.

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java [hunk 2]
```diff
@@ -95,23 +104,30 @@ protected MirrorTransaction verifyMirrorTransactionsResponse(MirrorNodeClient mi
...
+                                .getAcceptanceTestProperties()
+                                .getFeatureProperties()
+                                .getMaxContractFunctionGas(),
+                        contractResource.initialBalance == 0 ? null : Hbar.fromTinybars(contractResource.initialBalance),
+                        null);
+                ContractId contractId = verifyCreateContractNetworkResponse();
+
+                return new DeployedContract(fileId, contractId, compiledSolidityArtifact);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Comment (0xivanov) [inline hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java:128]: Might be better to catch all exceptions and log some additional info. For example                 
`log.warn("Issue creating contract: {}, ex: {}", contractResource, e);`

Comment (yiliev0) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java]: Done

File: hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java (hunk 1 above)
Comment (yiliev0) [reply hedera-mirror-test/src/test/java/com/hedera/mirror/test/e2e/acceptance/steps/AbstractFeature.java]: Implemented, thanks for the heads up.

**Comment by attest08** [comment]
LGTM

**Review by 0xivanov — APPROVED** [review]
LGTM

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Pull request merged**